### PR TITLE
feat: Extend Feature object with new attributes

### PR DESCRIPTION
**Pull request by akalex**
# Description

Currently, there is no way to get answers to the question - **Whether impression data is enabled?**

The good news is that this information is provided so we can store it and use it if needed.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] Unit tests
- [x] Spec Tests
- [ ] Integration tests / Manual Tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/specification_tests/test_client_specs.py [hunk 1]
```diff
@@ -29,6 +29,9 @@ def load_specs():
 def iter_spec():
     for spec in load_specs():
         name, state, tests, variant_tests = load_spec(spec)
+        state["features"] = [
+            {"impressionData": False, "stale": False, **feature} for feature in state["features"]
+        ]
```
Comment (akalex) [inline tests/specification_tests/test_client_specs.py:34]: I should be able to remove these lines once https://github.com/Unleash/client-specification/pull/51 has merged

File: UnleashClient/loader.py [hunk 2]
```diff
@@ -62,7 +62,9 @@ def _create_feature(provisioning: dict,
     return Feature(name=provisioning["name"],
                    enabled=provisioning["enabled"],
                    strategies=parsed_strategies,
-                   variants=variant
+                   variants=variant,
+                   impression_data=provisioning["impressionData"],
+                   stale=provisioning["stale"]
```
Comment (sighphyre) [inline UnleashClient/loader.py:67]: Looks like these will fail to load if these properties aren't present, which I'd rather avoid. `impression_data` won't be there prior to Unleash 4.7 and I'm not certain we can guarantee either of them in the bootstrap files. Can we swap this out for something more tolerant to these properties being absent like `provisioning.get("impressionData")`?

File: tests/specification_tests/test_client_specs.py (hunk 1 above)
Comment (sighphyre) [reply tests/specification_tests/test_client_specs.py:34]: Hopefully this won't be necessary if we address the previous comment

**Review by sighphyre — COMMENTED** [review]
Two thoughts here:
1) This seems to add the property for impression data but it doesn't actually add the functionality. Was that intended?

2) `stale` usually isn't actually used meaningfully by the SDKs, it seems like you're wanting to get your hands on this property within the SDK so I'm super curious, what's your usecase here? I'm not against exposing them but usually when I see this kind of pull request, the SDK is being used for something we didn't design them for

**Comment by akalex** [comment]
> stale usually isn't actually used meaningfully by the SDKs, it seems like you're wanting to get your hands on this property within the SDK so I'm super curious, what's your usecase here? I'm not against exposing them but usually when I see this kind of pull request, the SDK is being used for something we didn't design them for

The indention to open this PR mainly was issued by the absence of possibility in Python SDK to get impression data and have no clue when it will be added. So, I learned that extra information is available as part of the data pulled by the provisioning loader, but the object `Feature` does not preserve that information. 
Maybe I use SDK improperly, and you can advise what would be a better approach to get to know about Impression Data. Thank you in advance.

File: UnleashClient/loader.py (hunk 2 above)
Comment (akalex) [reply UnleashClient/loader.py:67]: Sure, I can make it fail-safe. I didn't know such details, so I thought that should be safe.

**Comment by sighphyre** [comment]
> > This seems to add the property for impression data but it doesn't actually add the functionality. Was that intended?
> 
> > stale usually isn't actually used meaningfully by the SDKs, it seems like you're wanting to get your hands on this property within the SDK so I'm super curious, what's your usecase here? I'm not against exposing them but usually when I see this kind of pull request, the SDK is being used for something we didn't design them for
> 
> The intention to open this PR mainly was issued by the absence of possibility in Python SDK to get impression data and have no clue when it will be added. So, I learned that extra information is available as part of the data pulled by the provisioning loader, but the object `Feature` does not preserve that information. Maybe I misuse SDK, and you can advise what would be a better approach to get to know about Impression Data. I really appreciate any help you can provide.

Right, this sounds like what you're doing here is starting the work/giving us a gentle nudge on implementing impression data in this SDK? That makes a lot of sense, that's absolutely not a misuse of this SDK. But this is a bigger feature than just implementing the properties that come back from the server. It's probably worth going going over the PRs that add this feature to the dotnet SDK (https://github.com/Unleash/unleash-client-dotnet/pull/113) and the node SDK (https://github.com/Unleash/unleash-client-node/pull/336) if you're interested in the technical details here and are looking to implement this here.

What probably needs to be done here is exposing the impression data property (done already), adding an event system or using Python built ins to expose a stream of events and then injecting the events into that system when the toggle is checked.

It sounds like this is also a request for us to implement this feature, so I've opened a ticket here for it: https://github.com/Unleash/unleash-client-python/issues/238 so that if this is more work than you had wanted to get involved we can track it internally.

**Comment by akalex** [comment]
> Right, this sounds like what you're doing here is starting the work/giving us a gentle nudge on implementing impression data in this SDK? That makes a lot of sense, that's absolutely not a misuse of this SDK. But this is a bigger feature than just implementing the properties that come back from the server. It's probably worth going going over the PRs that add this feature to the dotnet SDK (https://github.com/Unleash/unleash-client-dotnet/pull/113) and the node SDK (https://github.com/Unleash/unleash-client-node/pull/336) if you're interested in the technical details here and are looking to implement this here.

> What probably needs to be done here is exposing the impression data property (done already), adding an event system or using Python built ins to expose a stream of events and then injecting the events into that system when the toggle is checked.

> It sounds like this is also a request for us to implement this feature, so I've opened a ticket here for it: https://github.com/Unleash/unleash-client-python/issues/238 so that if this is more work than you had wanted to get involved we can track it internally.

You are correct that would be absolutely great if you could prioritise some work around Impression Data, and thanks for opening a ticket.

On the other hand, I still do not see a strong reason why the `Feature` object should not be extended with the properties that the object already has. Such changes neither break interface nor go against the rules but rather provides fair compromise and room for maneuvers.

What are your thoughts on that?

**Comment by sighphyre** [comment]
@akalex I'm very open to including the inclusion of of the impression data property before we've implemented the feature itself but I'm still not sure what including the stale property would add. Do you see a use case here that I'm missing?

**Comment by akalex** [comment]
> @akalex I'm very open to including the inclusion of of the impression data property before we've implemented the feature itself but I'm still not sure what including the stale property would add. Do you see a use case here that I'm missing?

@sighphyre that's a great question, I don't have a good example from the top of my head. Let me get rid of the `stale` property.

File: UnleashClient/loader.py (hunk 2 above)
Comment (akalex) [reply UnleashClient/loader.py]: this is done.

**Comment by akalex** [comment]
> > @akalex I'm very open to including the inclusion of of the impression data property before we've implemented the feature itself but I'm still not sure what including the stale property would add. Do you see a use case here that I'm missing?
> 
> @sighphyre that's a great question, I don't have a good example from the top of my head. Let me get rid of the `stale` property.

@sighphyre this is done. could you take a look at it once you have a chance

**Comment by sighphyre** [comment]
@akalex Thanks so much for doing the changes. This looks good to me. To keep you in the loop, I've raised the impression data feature with the team internally to look into next week, we'll probably just build on top of this PR

**Review by sighphyre — APPROVED** [review]
LGTM

**Comment by akalex** [comment]
> @akalex Thanks so much for doing the changes. This looks good to me. To keep you in the loop, I've raised the impression data feature with the team internally to look into next week, we'll probably just build on top of the work in this PR

@sighphyre thank you so much, I appreciate your help in pushing this one further. One last question if you don't mind, when these changes optimistically could be released?

**Comment by sighphyre** [comment]
Hey @akalex, probably when we finish off the feature, can't make promises until the team plans this but I will raise it for next week, I'm not sure there's a point in doing a release until this implements the feature fully, right now it's just a data property

**Comment by akalex** [comment]
> Hey @akalex, probably when we finish off the feature, can't make promises until the team plans this but I will raise it for next week, I'm not sure there's a point in doing a release until this implements the feature fully, right now it's just a data property

@sighphyre It's data property, but even without the Impression data feature, it allows us to build and use custom solutions like in my case. So I either have to fork and support that I don't really want to or you can release it as a new version.

**Pull request merged**